File: InterpreterInGo/go/lexer/lexer.go
package lexer

import "monkey/token"

type Lexer struct {
	input        []rune
	position     int
	readPosition int
	ch           rune
}

func New(input string) *Lexer {
	l := &Lexer{input: []rune(input)}
	l.readChar()
	return l
}

func (l *Lexer) NextToken() token.Token {
	var tok token.Token

	l.skipWhitespace()

	switch l.ch {
	case '=':
		if l.peekChar() == '=' {
			ch := l.ch
			l.readChar()
			tok = token.Token{Type: token.EQ, Literal: string(ch) + string(l.ch)}
		} else {
			tok = newToken(token.ASSIGN, l.ch)
		}
	case '+':
		tok = newToken(token.PLUS, l.ch)
	case '-':
		tok = newToken(token.MINUS, l.ch)
	case '/':
		tok = newToken(token.SLASH, l.ch)
	case '*':
		tok = newToken(token.STAR, l.ch)
	case '<':
		tok = newToken(token.LT, l.ch)
	case '>':
		tok = newToken(token.GT, l.ch)
	case '!':
		if l.peekChar() == '=' {
			ch := l.ch
			l.readChar()
			tok = token.Token{Type: token.NOT_EQ, Literal: string(ch) + string(l.ch)}
		} else {
			tok = newToken(token.BANG, l.ch)
		}
	case '(':
		tok = newToken(token.LPAREN, l.ch)
	case ')':
		tok = newToken(token.RPAREN, l.ch)
	case '{':
		tok = newToken(token.LBRACE, l.ch)
	case '}':
		tok = newToken(token.RBRACE, l.ch)
	case ',':
		tok = newToken(token.COMMA, l.ch)
	case ';':
		tok = newToken(token.SEMICOLON, l.ch)
	case 0:
		tok = token.Token{Type: token.EOF, Literal: ""}
	case '"':
		str := l.readString()
		tok = token.Token{Type: token.STRING, Literal: str}
	case '[':
		tok = newToken(token.LBRACKET, l.ch)
	case ']':
		tok = newToken(token.RBRACKET, l.ch)
	case ':':
		tok = newToken(token.COLON, l.ch)
	default:
		if isLetter(l.ch) {
			literal := l.readIdentifier()
			tok.Type = token.LookupToken(literal)
			tok.Literal = literal
			return tok
		} else if isNumerical(l.ch) {
			number := l.readNumber()
			tok.Type = token.INT
			tok.Literal = number
			return tok
		} else {
			tok = newToken(token.ILLEGAL, l.ch)
		}
	}
	l.readChar()
	return tok
}

func newToken(tokType token.TokenType, literal rune) token.Token {
	return token.Token{Type: tokType, Literal: string(literal)}
}

func (l *Lexer) readString() string {
	pos := l.position + 1
	for {
		l.readChar()
		if l.ch == '"' || l.ch == 0 {
			break
		}
	}

	return string(l.input[pos:l.position])

}

func (l *Lexer) readIdentifier() string {
	pos := l.position
	for isLetter(l.ch) {
		l.readChar()
	}
	return string(l.input[pos:l.position])
}

func (l *Lexer) readNumber() string {
	pos := l.position
	for isNumerical(l.ch) {
		l.readChar()
	}
	return string(l.input[pos:l.position])
}

func (l *Lexer) readChar() {
	if l.readPosition >= len(l.input) {
		l.ch = 0
	} else {
		l.ch = l.input[l.readPosition]
	}
	l.position = l.readPosition
	l.readPosition++
}

func (l *Lexer) peekChar() rune {
	if l.readPosition >= len(l.input) {
		return 0
	}
	return l.input[l.readPosition]
}

func (l *Lexer) skipWhitespace() {
	for isWhitespace(l.ch) {
		l.readChar()
	}
}

func isWhitespace(r rune) bool {
	return r == ' ' || r == '\t' || r == '\n' || r == '\r'
}

func isLetter(r rune) bool {
	return ('a' <= r && r <= 'z') || ('A' <= r && r <= 'Z') || r == '_'
}

func isNumerical(r rune) bool {
	return '0' <= r && r <= '9'
}

File: InterpreterInGo/dart/lib/lexer.dart
import 'package:monkey_intepreter/constants.dart';
import 'package:monkey_intepreter/token.dart';

class Lexer {
  final RegExp _alphaRegex = RegExp(r'^[a-zA-Z]$');
  final String _input;

  int _readPosition = 0;

  Lexer(this._input);

  Token nextToken() {
    if (_readPosition >= _input.length) return EndOfFile();

    _skipWhitespace();

    if (_readPosition >= _input.length) return EndOfFile();

    Token? token;

    switch (_currentChar()) {
      case "+":
        token = Plus();
        break;
      case '-':
        token = Minus();
        break;
      case '*':
        token = Star();
        break;
      case '/':
        token = Slash();
        break;
      case '{':
        token = LeftBrace();
        break;
      case '}':
        token = RightBrace();
        break;
      case '(':
        token = LeftParen();
        break;
      case ')':
        token = RightParen();
        break;
      case '[':
        token = LeftBracket();
        break;
      case ']':
        token = RightBracket();
        break;
      case ',':
        token = Comma();
        break;
      case ':':
        token = Colon();
        break;
      case ';':
        token = SemiColon();
        break;
      case '=':
        if (_nextChar() == "=") {
          _readPosition += 1;
          token = Equals();
          break;
        }
        token = Assign();
        break;
      case '!':
        if (_nextChar() == "=") {
          _readPosition += 1;
          token = NotEquals();
          break;
        }
        token = Bang();
        break;
      default:
        if (_isAlpha(_currentChar())) {
          var literal = _readIdentifier();
          var keywordFunc = _keywords[literal];
          return keywordFunc == null
              ? Identifier(literal)
              : keywordFunc(literal);
        }

        if (_isNumber(_currentChar())) {
          token = _readNumber();
        }
    }

    _readPosition += 1;

    return token ?? Illegal();
  }

  String? _currentChar() =>
      _readPosition >= _input.length ? null : _input[_readPosition];

  String? _nextChar() =>
      _readPosition + 1 >= _input.length ? null : _input[_readPosition + 1];

  bool _isNumber(String? c) => c == null ? false : double.tryParse(c) != null;
  bool _isAlpha(String? c) => c == null ? false : _alphaRegex.hasMatch(c);

  void _skipWhitespace() {
    var currChar = _input[_readPosition];
    while (Constants.whitespace.contains(currChar) &&
        _readPosition < _input.length) {
      _readPosition++;
      if (_readPosition >= _input.length) break;
      currChar = _input[_readPosition];
    }
  }

  String _readIdentifier() {
    var startPosition = _readPosition;
    while (_isAlpha(_currentChar())) {
      _readPosition += 1;
    }
    return _input.substring(startPosition, _readPosition);
  }

  Token _readNumber() {
    var firstPart = _parseNumeric();
    if (_currentChar() != ".") return Integer(firstPart);
    _readPosition += 1;
    var secondPart = _parseNumeric();
    return Float("$firstPart.$secondPart");
  }

  String _parseNumeric() {
    var startPosition = _readPosition;
    while (_isNumber(_currentChar())) {
      _readPosition += 1;
    }
    return _input.substring(startPosition, _readPosition);
  }

  final Map<String, Token Function(String)> _keywords = {
    "func": Func.new,
    "let": Let.new,
    "if": If.new,
    "return": Return.new
  };
}
